model/riscv_insts_vext_utils.sail

```diff
@@ -0,0 +1,1165 @@
+
+  /* Determine the actual number of elements when lmul < 1 */
+  let real_num_elem = if LMUL_pow >= 0 then num_elem else num_elem / int_power(2, 0 - LMUL_pow);
+  assert(num_elem >= real_num_elem);
+
+  foreach (i from 0 to (num_elem - 1)) {
+    if i < start_element then {
+      /* Prestart elements defined by vstart */
+      result[i] = vd_val[i];
+      mask[i] = false
+    } else if i > end_element then {
+      /* Tail elements defined by vl */
+      /* Mask tail is always agnostic */
+      result[i] = vd_val[i]; /* TODO: configuration support */
+      mask[i] = false
+    } else if i >= real_num_elem then {
+      /* Tail elements defined by lmul < 1 */
+      /* Mask tail is always agnostic */
+      result[i] = vd_val[i]; /* TODO: configuration support */
+      mask[i] = false
+    } else if not(vm_val[i]) then {
+      /* Inactive body elements defined by vm */
+      result[i] = match mask_ag {
+        UNDISTURBED => vd_val[i],
+        AGNOSTIC    => vd_val[i] /* TODO: configuration support */
+      };
+      mask[i] = false
+    } else {
+      /* Active body elements */
+      mask[i] = true
+    }
+  };
+
+  (result, mask)
+}
+
+/* For vector load/store segment instructions:
+ *   Read multiple register groups and concatenate them in parallel
+ *   The whole segments with the same element index are combined together
+ */
+val read_vreg_seg : forall 'n 'm 'p 'q. (int('n), int('m), int('p), int('q), regidx) -> vector('n, dec, bits('q * 'm)) effect {escape, rreg, undef}
+function read_vreg_seg(num_elem, SEW, LMUL_pow, nf, vrid) = {
+  assert('q * 'm > 0);
+  let LMUL_reg : int = if LMUL_pow <= 0 then 1 else int_power(2, LMUL_pow);
+  vreg_list : vector('q, dec, vector('n, dec, bits('m))) = undefined;
+  result : vector('n, dec, bits('q * 'm)) = undefined;
+  foreach (j from 0 to (nf - 1)) {
+    vreg_list[j] = read_vreg(num_elem, SEW, LMUL_pow, vrid + to_bits(5, j * LMUL_reg));
+  };
+  foreach (i from 0 to (num_elem - 1)) {
+    result[i] = zeros('q * 'm);
+    foreach (j from 0 to (nf - 1)) {
+      result[i] = result[i] | (zero_extend(vreg_list[j][i]) << (j * 'm))
+    }
+  };
+  result
+}
+
+/* Floating point canonical NaN for 16-bit, 32-bit and 64-bit types */
+val canonical_NaN : forall 'm, 'm in {16, 32, 64}. int('m) -> bits('m)
+function canonical_NaN('m) = {
+  match 'm {
+    16 => canonical_NaN_H(),
+    32 => canonical_NaN_S(),
+    64 => canonical_NaN_D()
+  }
+}
+
+/* Floating point NaN boxing / unboxing that support 16-bit to 64-bit types */
+val nan_box_vector : forall 'm, ('m in {16, 32, 64} & 'm <= flen). bits('m) -> flenbits
+function nan_box_vector(unboxed) = {
+  match 'm {
+    16 => nan_box(unboxed),
+    32 => nan_box(unboxed),
+    64 => unboxed
+  }
+}
+
+val nan_unbox_vector : forall 'm, ('m in {16, 32, 64} & 'm <= flen). (flenbits, int('m)) -> bits('m)
+function nan_unbox_vector(regval, 'm) = {
+  match 'm {
+    16 => nan_unbox(regval),
+    32 => nan_unbox(regval),
+    64 => regval
+  }
+}
```

Comment:
These can go once you change each of their sole use to use F_H/S/D instead

Reply:
nan_box and unbox only have H and S, without D, so I have to write this to make sure it can take 16, 32 and 64 bit input.

Reply:
Got it. I'll modify this part.

Reply:
These functions are deleted.